File: Sources/Transport/ReplayProtection.swift
import Foundation
// filename: NonceCache.swift

import Primitives

/// Cache for tracking nonces to prevent replay attacks.
///
/// Stores nonces with timestamps and automatically expires old entries.
/// Used to ensure each packet is processed only once.
public actor ReplayProtection {
    private var seenNonces: [Data: Date] = [:]
    private let maxAge: TimeInterval

    /// Creates a nonce cache with specified max age.
    ///
    /// - Parameter maxAge: How long nonces are remembered (default: 300 seconds / 5 minutes)
    public init(maxAge: TimeInterval = 300.0) {
        self.maxAge = maxAge
    }

    /// Checks if a nonce has been seen before.
    ///
    /// - Parameter nonce: The nonce to check
    /// - Returns: true if already seen, false if new
    public func hasSeen(_ nonce: Data) -> Bool {
        cleanup()
        return seenNonces[nonce] != nil
    }

    /// Records a nonce as seen.
    ///
    /// - Parameter nonce: The nonce to record
    public func record(_ nonce: Data) {
        seenNonces[nonce] = Date()

        // Cleanup after every 100 records to prevent unbounded growth
        if seenNonces.count % 100 == 0 {
            cleanup()
        }
    }

    /// Checks and records a nonce in one operation.
    ///
    /// - Parameter nonce: The nonce to check and record
    /// - Returns: true if this is a new nonce (allowed), false if already seen (replay)
    public func checkAndRecord(_ nonce: Data) -> Bool {
        if hasSeen(nonce) {
            return false  // Replay detected
        }
        record(nonce)
        return true  // New nonce, allowed
    }

    /// Returns the number of nonces currently cached.
    public var count: Int {
        seenNonces.count
    }

    /// Clears all cached nonces.
    public func clear() {
        seenNonces.removeAll()
    }

    /// Removes nonces older than maxAge.
    private func cleanup() {
        let now = Date()
        seenNonces = seenNonces.filter { _, timestamp in
            now.timeIntervalSince(timestamp) < maxAge
        }
    }
}

/// Extension for packet timestamp validation
extension Date {
    /// Checks if this timestamp is within acceptable age.
    ///
    /// - Parameter maxAge: Maximum allowed age in seconds
    /// - Returns: true if timestamp is recent enough
    public func isRecent(maxAge: TimeInterval) -> Bool {
        let now = Date()
        let age = now.timeIntervalSince(self)
        return age >= 0 && age <= maxAge
    }
}

File: src/python/yx/transport/replay_protection.py
"""
Replay protection for YX protocol.

Prevents replay attacks by tracking nonces (HMACs) with timestamps.
Automatically expires old nonces to prevent memory growth.

Implementation matches Swift ReplayProtection actor.
"""

import time
from typing import Dict


class ReplayProtection:
    """Cache for tracking nonces to prevent replay attacks"""

    def __init__(self, max_age: float = 300.0):
        """
        Creates a nonce cache with specified max age.

        Args:
            max_age: How long nonces are remembered (default: 300 seconds / 5 minutes)
        """
        self._seen_nonces: Dict[bytes, float] = {}
        self._max_age = max_age

    def has_seen(self, nonce: bytes) -> bool:
        """Check if nonce has been seen before"""
        self._cleanup()
        return nonce in self._seen_nonces

    def record(self, nonce: bytes) -> None:
        """Record a nonce as seen"""
        self._seen_nonces[nonce] = time.time()

        # Cleanup after every 100 records
        if len(self._seen_nonces) % 100 == 0:
            self._cleanup()

    def check_and_record(self, nonce: bytes) -> bool:
        """
        Check and record nonce in one operation.

        Returns:
            True if new nonce (allowed), False if already seen (replay)
        """
        if self.has_seen(nonce):
            return False  # Replay detected
        self.record(nonce)
        return True  # New nonce, allowed

    @property
    def count(self) -> int:
        """Returns number of nonces currently cached"""
        return len(self._seen_nonces)

    def clear(self) -> None:
        """Clear all cached nonces"""
        self._seen_nonces.clear()

    def _cleanup(self) -> None:
        """Remove nonces older than max_age"""
        now = time.time()
        self._seen_nonces = {
            nonce: timestamp
            for nonce, timestamp in self._seen_nonces.items()
            if now - timestamp < self._max_age
        }
